build.gradle

```
@@ -582,6 +582,154 @@ task mapNamedJar(dependsOn: ["mergeV2", mapIntermediaryJar]) {
 	}
 }
 
+import daomephsta.unpick.api.ConstantUninliner
+import daomephsta.unpick.api.constantmappers.ConstantMappers
+import daomephsta.unpick.api.constantresolvers.ConstantResolvers
+import daomephsta.unpick.api.IClassResolver
+import org.objectweb.asm.ClassReader
+import java.util.jar.JarEntry
+import java.util.jar.JarFile
+import java.util.jar.JarOutputStream
+import org.objectweb.asm.tree.ClassNode
+import org.objectweb.asm.ClassWriter
+
+class JarClassResolver implements IClassResolver, Closeable {
+	private final URLClassLoader classLoader;
+
+	JarClassResolver(Set<File> jars) {
+		this.classLoader = new URLClassLoader(
+			jars.stream().map{file -> file.toURI().toURL()}.toArray{new URL[it]}
+		)
+	}
+
+	JarClassResolver(File jar) {
+		this(Collections.singleton(jar))
+	}
+
+	@Override
+	ClassReader resolveClass(String internalName) throws IOException {
+		def resource = classLoader.getResourceAsStream(internalName + ".class")
+
+		if (resource != null) {
+			return new ClassReader(resource)
+		}
+
+		throw new RuntimeException("Failed to find ${internalName}.class")
+	}
+
+	@Override
+	void close() throws IOException {
+		classLoader.close()
+	}
+}
+
+// Merge all of the unpick definitions into a single file to be shipped
+task combineUnpickDefinitions(type : FileOutput) {
+	output = new File(tempDir, "definitions.unpick")
+	outputs.upToDateWhen { false }
+	group "unpick"
+
+	doFirst {
+		output.delete()
+	}
+
+	doLast {
+		def combinedDefinitions = "v2\n"
+
+		file("unpick-definitions").eachFile {
+			def definition = it.text
+			if (!definition.startsWith("v2")) {
+				throw new UnsupportedOperationException(it.name + " has an invalid unpick header")
+			}
+
+			// Remove the first, and append to the combined definition
+			combinedDefinitions += definition.split("\n").drop(1).join("\n")
+		}
+
+		output.text = combinedDefinitions
```

If desired, syntax checking can be performed here using unpick-format-utils reader and writer classes. 
`UnpickV2Writer` doesn't visit the version, so stripping it isn't necessary. 
```suggestion
		def combinedDefinitions = new UnpickV2Writer()

		file("unpick-definitions").eachFile { file ->
			file.withInputStream { is -> 
				byte[] version = new byte [2];
				is.read(version);
				if (version != ['v', '2'])
					throw new UnsupportedOperationException(file.name + " has an invalid unpick header")
				// Reader does not need closing, because is will be closed at the closure end
				new UnpickV2Reader(is).accept(combinedDefinitions) 
			}
		}

		output.text = combinedDefinitions.output
```

---

That sounds like a much better idea than what I have here. I’ll take a look 👍

---

Made the changes, its much better thanks.